src/platform/monitoring/DowntimeNotification/util/helpers.js

```diff
@@ -0,0 +1,76 @@
+import moment from 'moment';
+import serviceStatus from '../config/serviceStatus';
+
+/**
+ * Derives downtime status based on a time range
+ * @param {string|Date|Moment} startTime
+ * @param {string|Date|Moment} endTime
+ * @returns {string} A service status
+ */
+export function getStatusForTimeframe(startTime, endTime) {
+  const now = moment();
+  const hasStarted = now.isSameOrAfter(startTime);
+
+  if (hasStarted) {
+    // Check for indefinite downtime (null endTime) or that the endTime is in the future
+    if (!endTime || now.isBefore(endTime)) {
+      return serviceStatus.down;
+    }
+    // The downtime must be old and outdated. The API should filter these so this shouldn't happen.
+    return serviceStatus.ok;
+  }
+
+  const startsWithinHour = now.add(1, 'hour').isSameOrAfter(startTime);
+  if (startsWithinHour) return serviceStatus.downtimeApproaching;
+
+  return serviceStatus.ok;
+}
+
+/**
+ * Creates a Map of downtime information using the "externalService" property as keys
+ * @param {Array} maintenanceWindows The raw JSON data from the API
+ * @returns {Map}
+ */
+export function createServiceMap(maintenanceWindows = []) {
+  const serviceMap = new Map();
+
+  for (const maintenanceWindow of maintenanceWindows) {
+    const {
+      attributes: {
+        externalService,
+        startTime: startTimeRaw,
+        endTime: endTimeRaw
+      }
+    } = maintenanceWindow;
+
+    const startTime = moment(startTimeRaw);
+    const endTime = endTimeRaw && moment(endTimeRaw);
+    const status = getStatusForTimeframe(startTime, endTime);
+
+    serviceMap.set(externalService, {
+      externalService,
+      status,
+      startTime,
+      endTime
+    });
+  }
+
+  return serviceMap;
+}
```

Comment:
This serviceMap will become available in the Redux store like -

```
const mapStateToProps = (state) => {
  return {
    evssDowntime: state.scheduledDowntime.isReady && state.serviceMap.get('evss')
  }
}
```
`this.props.evssDowntime` would have the shape -

```
{
  externalService: string // 'evss'
  status: string // down, downtimeApproaching, or ok
  startTime: moment,
  endTime: moment  // may be undefined if indefinite downtime
}
```
It may also be null if there isn't any scheduled downtime.

Reply:
What does `isReady` mean?

Reply:
It means that the request for scheduled downtime is finished and the component/underlying data is ready for render.